From e5264cdca9e8aaefcd1aad81e69d3edad036dcdf Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Fri, 10 Jun 2022 05:04:22 +0300
Subject: [PATCH] [SW-89971] rom: add dma dest type flexibility

Based on the destination, the dma dest type can either be hbw or sram.
This patch adds the flexibility based on load address. This is in
preparation to reuse the image bringup function for tpm image.

Change-Id: I1dfb7fe97a9bfbeb5f69144325ddd7c8660c9ab0
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 9a3b548..4864f99 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -542,6 +542,7 @@
 	enum arcboot_rom_progress img_sts_base = PROGRESS_BRINGUP_PCI_PRI;
 	enum dma_error err = DMA_SUCCESS;
 	int ret;
+	enum dma_dest_type dst_type;
 
 	arcboot_rom_global.curr_img = cur_img;
 
@@ -563,8 +564,10 @@
 		return ARCBOOT_BRINGUP_ERROR;
 	}
 
+	dst_type = arcboot_rom_get_dma_dest_type(desc->loadaddr);
+
 	err = dma_copy_image_v2(desc->img_size, desc->spi_offset,
-				desc->loadaddr, DMA_DEST_SRAM, false);
+				desc->loadaddr, dst_type, false);
 	if (err != DMA_SUCCESS) {
 		if (err == DMA_INVALID_CRC_ERROR) {
 			security_report_img_status(cur_img, IMG_STATUS_CRC_FAIL);
diff --git a/arch/gaudi3/arc/dev_common.h b/arch/gaudi3/arc/dev_common.h
index abb2a27..27ea0a5 100644
--- a/arch/gaudi3/arc/dev_common.h
+++ b/arch/gaudi3/arc/dev_common.h
@@ -30,3 +30,4 @@
 int arcboot_rom_fuse_single_die_config(void);
 void disable_spi_gw(void);
 uint64_t get_pcie_pll_freq(enum pll_div_out div);
+enum dma_dest_type arcboot_rom_get_dma_dest_type(uint64_t addr);
diff --git a/arch/gaudi3/arc/recovery.c b/arch/gaudi3/arc/recovery.c
index cd06706..bfe7b6a 100644
--- a/arch/gaudi3/arc/recovery.c
+++ b/arch/gaudi3/arc/recovery.c
@@ -16,7 +16,7 @@
 #include <dev_efuse.h>
 #include <dev_dma.h>
 
-static enum dma_dest_type arcboot_rom_get_dma_dest_type(uint64_t addr)
+enum dma_dest_type arcboot_rom_get_dma_dest_type(uint64_t addr)
 {
 	if (addr >= HL_FW_SRAM_FULL_BASE &&
 	    addr < (HL_FW_SRAM_FULL_BASE + HL_FW_SRAM_SIZE))
